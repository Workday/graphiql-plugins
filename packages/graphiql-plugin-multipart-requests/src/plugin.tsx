import { GraphiQLPlugin, PenIcon, TrashIcon, useGraphiQL } from '@graphiql/react';
import React, { CSSProperties, useContext, useEffect, useRef, useState } from 'react';
import uploadClipSvg from './attachment-icon.svg.js';
import { AttachmentDeleteAction, AttachmentUpdateAction, MultipartAttachmentsInternalContext } from './context.js';

import '../../style.css';

type AttachmentContentProps = {
  name: string;
  value: File | null;
  updateAttachment: (action: Omit<AttachmentUpdateAction, 'id'> | Omit<AttachmentDeleteAction, 'id'>) => void;
  error?: string;
};

const GRAPHIQL_PLUGIN: GraphiQLPlugin = {
  title: 'Multipart Attachments',
  icon: () => {
    const [attachments] = useContext(MultipartAttachmentsInternalContext);
    const visiblePlugin = useGraphiQL(state => state.visiblePlugin);

    const style: CSSProperties = {
      fontFamily: 'var(--font-family-mono)',
      fontSize: 'var(--font-size-inline-code)',
      textAlign: 'center',
      width: 'auto',
    };
    if (visiblePlugin === GRAPHIQL_PLUGIN) {
      style.fontWeight = 'bold';
    }

    return (
      <div className='attachments-icon'>
        {uploadClipSvg}
        {attachments.validAttachments.length > 0 ? (
          <span className='attachments-icon-badge'>{attachments.validAttachments.length}</span>
        ) : (
          <></>
        )}
      </div>
    );
  },
  content: () => {
    const [attachments, attachmentsDispatch] = useContext(MultipartAttachmentsInternalContext);

    return (
      <div className={'attachments-container'}>
        <h2>Multipart Attachments</h2>
        {attachments.map(({ id, name, value, error }) => {
          return (
            <AttachmentContent
              key={id}
              name={name}
              value={value}
              updateAttachment={(action) => attachmentsDispatch({ ...action, id })}
              error={error}
            />
          );
        })}
        <div style={{ textAlign: 'center' }}>
          <button onClick={() => attachmentsDispatch({ action: 'add' })}>Add Part</button>
        </div>
      </div>
    );
  },
};

export function AttachmentContent({ name, value, updateAttachment, error }: AttachmentContentProps): React.JSX.Element {
  const fileInputRef = useRef<HTMLInputElement | null>(null);

  return (
    <div className={'attachment-content'}>
      <div className={'attachment-title'}>
        <PartNameComponent name={name} setName={(name) => updateAttachment({ action: 'update', name })} />
        <button onClick={() => updateAttachment({ action: 'delete' })} className={'attachment-delete'}>
          <TrashIcon />
        </button>
      </div>
      {error ? <div className='graphiql-doc-explorer-error'>{error}</div> : <></>}
      {value ? <PartHeadersComponent name={name} value={value} /> : <></>}
      <div>
        <button style={{ marginTop: '0.5em' }} onClick={() => fileInputRef.current?.click()}>
          Attach File
        </button>
        <input
          type='file'
          ref={fileInputRef}
          onChange={(event) => {
            const file = event.target.files?.[0];
            if (file) {
              updateAttachment({ action: 'update', value: file });
            }
          }}
          style={{ display: 'none' }}
        />
      </div>
    </div>
  );
}

export function PartNameComponent({ name, setName }: { name: string; setName: (name: string) => void }) {
  const [isEditing, setIsEditing] = useState(false);
  const inputRef = useRef<HTMLInputElement | null>(null);

  useEffect(() => {
    if (isEditing) {
      inputRef.current?.focus();
    }
  }, [isEditing]);

  function endEditing() {
    setIsEditing(false);
    if (inputRef.current) {
      setName(inputRef.current.value);
    }
  }
  function onSubmit(event: React.BaseSyntheticEvent) {
    event.preventDefault();
    event.stopPropagation();
    endEditing();
  }

  if (isEditing) {
    return (
      <form onSubmit={onSubmit}>
        <input
          ref={inputRef}
          type='text'
          defaultValue={name}
          onBlur={endEditing}
          onKeyDown={(e) => {
            if (e.key === 'Escape') {
              onSubmit(e);
            }
          }}
        />
      </form>
    );
  } else {
    return (
      <>
        <span onClick={() => setIsEditing(true)} style={{ cursor: 'pointer' }}>
          {name}
          &nbsp;
          <PenIcon />
        </span>
      </>
    );
  }
}

export function PartHeadersComponent({ name, value }: { name: string; value: File }) {
  const baseContentDisposition = `form-data; name="${encodeURIComponent(name)}"`;
  let headers;
  if (typeof value !== 'string') {
    headers = [
      ['Content-Disposition', `${baseContentDisposition}; filename="${encodeURIComponent(value.name)}"`],
      ['Content-Type', value.type.length > 0 ? value.type : 'application/octet-stream'],
    ];
  } else {
    headers = [['Content-Disposition', baseContentDisposition]];
  }

  return (
    <div className={'attachment-header'}>
      {headers.map(([name, value]) => (
        <div key={name}>
          <span style={{ color: 'hsl(var(--color-info))' }}>{name}: </span>
          <span>{value}</span>
        </div>
      ))}
    </div>
  );
}

export default GRAPHIQL_PLUGIN;
