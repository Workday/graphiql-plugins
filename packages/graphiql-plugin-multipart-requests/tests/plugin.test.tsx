/** @jest-environment jsdom */

import { AttachmentContent, PartHeadersComponent, PartNameComponent } from '../src/plugin.js';
import { act, render } from '@testing-library/react';

jest.mock('@graphiql/react', () => {
  return {
    PenIcon: () => <div>PenIcon</div>,
    TrashIcon: () => <div>PenIcon</div>,
    usePluginContext: () => null,
  };
});

describe('AttachmentContent', () => {
  test('renders a valid attachment', () => {
    expect(
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      render(<AttachmentContent name='a' value={new File([], 'A')} updateAttachment={() => {}} />).container
    ).toMatchSnapshot();
  });
  test('renders an invalid attachment', () => {
    expect(
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      render(<AttachmentContent name='a' value={new File([], 'A')} error={'File Error!'} updateAttachment={() => {}} />)
        .container
    ).toMatchSnapshot();
  });
});

describe('PartNameComponent', () => {
  test('renders a part', () => {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    expect(render(<PartNameComponent name='a' setName={() => {}} />).container).toMatchSnapshot();
  });

  test('updates a name', () => {
    const setName = jest.fn();
    const container = document.body.appendChild(document.createElement('div'));
    act(() => {
      return render(<PartNameComponent name='a' setName={setName} />, { container });
    });
    act(() => {
      container.querySelector('span')!.click();
    });
    act(() => {
      const input = container.querySelector('input')!;
      input.value = 'newName';
      input.blur();
    });
    expect(setName).toHaveBeenCalledWith('newName');
  });
});
describe('PartHeadersComponent', () => {
  test('renders a file', () => {
    expect(
      render(<PartHeadersComponent name='a' value={new File([], 'A', { type: 'text/plain' })} />).container
    ).toMatchSnapshot();
  });
  test('defaults to octet-stream', () => {
    expect(render(<PartHeadersComponent name='a' value={new File([], 'A')} />).container).toMatchSnapshot();
  });
});
