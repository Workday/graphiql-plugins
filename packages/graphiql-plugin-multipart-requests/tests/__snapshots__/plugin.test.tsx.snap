// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentContent renders a valid attachment 1`] = `
<div>
  <div
    class="attachment-content"
  >
    <div
      class="attachment-title"
    >
      <span
        style="cursor: pointer;"
      >
        a
         
        <div>
          PenIcon
        </div>
      </span>
      <button
        class="attachment-delete"
      >
        <div>
          PenIcon
        </div>
      </button>
    </div>
    <div
      class="attachment-header"
    >
      <div>
        <span>
          Content-Disposition
          : 
        </span>
        <span>
          form-data; name="a"; filename="A"
        </span>
      </div>
      <div>
        <span>
          Content-Type
          : 
        </span>
        <span>
          application/octet-stream
        </span>
      </div>
    </div>
    <div>
      <button
        style="margin-top: 0.5em;"
      >
        Attach File
      </button>
      <input
        style="display: none;"
        type="file"
      />
    </div>
  </div>
</div>
`;

exports[`AttachmentContent renders an invalid attachment 1`] = `
<div>
  <div
    class="attachment-content"
  >
    <div
      class="attachment-title"
    >
      <span
        style="cursor: pointer;"
      >
        a
         
        <div>
          PenIcon
        </div>
      </span>
      <button
        class="attachment-delete"
      >
        <div>
          PenIcon
        </div>
      </button>
    </div>
    <div
      class="graphiql-doc-explorer-error"
    >
      File Error!
    </div>
    <div
      class="attachment-header"
    >
      <div>
        <span>
          Content-Disposition
          : 
        </span>
        <span>
          form-data; name="a"; filename="A"
        </span>
      </div>
      <div>
        <span>
          Content-Type
          : 
        </span>
        <span>
          application/octet-stream
        </span>
      </div>
    </div>
    <div>
      <button
        style="margin-top: 0.5em;"
      >
        Attach File
      </button>
      <input
        style="display: none;"
        type="file"
      />
    </div>
  </div>
</div>
`;

exports[`PartHeadersComponent defaults to octet-stream 1`] = `
<div>
  <div
    class="attachment-header"
  >
    <div>
      <span>
        Content-Disposition
        : 
      </span>
      <span>
        form-data; name="a"; filename="A"
      </span>
    </div>
    <div>
      <span>
        Content-Type
        : 
      </span>
      <span>
        application/octet-stream
      </span>
    </div>
  </div>
</div>
`;

exports[`PartHeadersComponent renders a file 1`] = `
<div>
  <div
    class="attachment-header"
  >
    <div>
      <span>
        Content-Disposition
        : 
      </span>
      <span>
        form-data; name="a"; filename="A"
      </span>
    </div>
    <div>
      <span>
        Content-Type
        : 
      </span>
      <span>
        text/plain
      </span>
    </div>
  </div>
</div>
`;

exports[`PartNameComponent renders a part 1`] = `
<div>
  <span
    style="cursor: pointer;"
  >
    a
     
    <div>
      PenIcon
    </div>
  </span>
</div>
`;
